eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java

```diff
@@ -2307,6 +2308,89 @@ private <P> V slowUpdateValueWith(
         }
     }
 
+    @Override
+    public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction)
+    {
+        Objects.requireNonNull(value, "value cannot be null");
+        Objects.requireNonNull(remappingFunction, "remappingFunction cannot be null");
+
+        int hash = this.hash(key);
+        AtomicReferenceArray currentArray = this.table;
+        int length = currentArray.length();
+        int index = ConcurrentHashMap.indexFor(hash, length);
+        Object o = currentArray.get(index);
+        if (o == null)
+        {
+            Entry<K, V> newEntry = new Entry<>(key, value, null);
+            if (currentArray.compareAndSet(index, null, newEntry))
+            {
+                this.addToSize(1);
+                return value;
+            }
+        }
+        return this.slowMerge(key, value, remappingFunction, hash, currentArray);
+    }
+
+    private V slowMerge(
+            K key,
+            V value,
+            BiFunction<? super V, ? super V, ? extends V> remappingFunction,
+            int hash,
+            AtomicReferenceArray currentArray)
+    {
+        //noinspection LabeledStatement
+        outer:
+        while (true)
+        {
+            int length = currentArray.length();
+            int index = ConcurrentHashMap.indexFor(hash, length);
+            Object o = currentArray.get(index);
+            if (o == RESIZED || o == RESIZING)
+            {
+                currentArray = this.helpWithResizeWhileCurrentIndex(currentArray, index);
+            }
+            else
+            {
+                Entry<K, V> e = (Entry<K, V>) o;
+                while (e != null)
+                {
+                    Object candidate = e.getKey();
+                    if (candidate.equals(key))
+                    {
+                        V oldValue = e.getValue();
+                        V newValue = remappingFunction.apply(oldValue, value);
+                        Entry<K, V> replacementChainForRemoval =
+                                this.createReplacementChainForRemoval((Entry<K, V>) o, e);
+                        Entry<K, V> newEntry = newValue == null
+                                ? replacementChainForRemoval
+                                : new Entry<>(
+                                        e.getKey(),
+                                        newValue,
+                                        replacementChainForRemoval);
+                        if (!currentArray.compareAndSet(index, o, newEntry))
+                        {
+                            //noinspection ContinueStatementWithLabel
+                            continue outer;
+                        }
+
+                        if (newValue == null)
+                        {
+                            this.addToSize(-1);
+                        }
+                        return newValue;
```

Comment:
This is the trickiest part of the implementation and deserves a careful look. If the lambda returns null, the contract says to remove the existing entry for the key. Unlike the updateValue and getIfAbesent methods, the map can grow, stay the same, or shrink from this method call.